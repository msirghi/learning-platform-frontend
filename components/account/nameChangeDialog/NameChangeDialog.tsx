import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  TextField
} from '@material-ui/core';
import React, { useEffect, useState } from 'react';
import { useTranslation } from '../../../i18n';
import styles from './NameChangeDialog.module.scss';

type Props = {
  open: boolean;
  toggleOpen: (val: boolean) => void;
  onSubmit: (firstName: string, lastName: string) => void;
};

export const NameChangeDialog: React.FC<Props> = ({ open, toggleOpen, onSubmit }) => {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const { t } = useTranslation();

  useEffect(() => {
    setFirstName('');
    setLastName('');
  }, [open]);

  return (
    <Dialog open={open} onClose={() => toggleOpen(false)}>
      <DialogTitle>{t('account:nameChangeDialogTitle')}</DialogTitle>
      <DialogContent>
        <DialogContentText>{t('account:nameChangeDialogDescription')}</DialogContentText>
        <TextField
          label={t('account:firstName')}
          fullWidth
          variant='outlined'
          value={firstName}
          inputProps={{
            'data-testid': 'firstname-field'
          }}
          onChange={(e) => setFirstName(e.target.value)}
        />
        <div className={styles.lastNameField}>
          <TextField
            label={t('account:lastName')}
            fullWidth
            variant='outlined'
            value={lastName}
            inputProps={{
              'data-testid': 'lastname-field'
            }}
            onChange={(e) => setLastName(e.target.value)}
          />
        </div>
      </DialogContent>
      <DialogActions>
        <Button onClick={() => toggleOpen(false)} color='primary'>
          {t('account:cancelButtonLabel')}
        </Button>
        <Button
          data-testid='submit-button'
          onClick={() => onSubmit(firstName, lastName)}
          color='primary'
          disabled={!firstName || !lastName}
        >
          {t('account:submitButtonLabel')}
        </Button>
      </DialogActions>
    </Dialog>
  );
};
