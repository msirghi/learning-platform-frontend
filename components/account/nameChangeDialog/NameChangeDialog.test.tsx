import { Button, Dialog, TextField } from '@material-ui/core';
import { fireEvent, render } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { NameChangeDialog } from './NameChangeDialog';

describe('NameChangeDialog component', () => {
  const props = {
    open: true,
    toggleOpen: jest.fn(),
    onSubmit: jest.fn((firstName: string, lastName: string) => {})
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should toggle the dialog status on cancel button click', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    const cancelButton = wrapper.find(Button).at(0);

    expect(cancelButton).toBeDefined();
    cancelButton.simulate('click');
    expect(props.toggleOpen).toBeCalled();
  });

  it('should toggle the dialog status on onClose Dialog prop', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    const dialog = wrapper.find(Dialog);

    expect(dialog).toBeDefined();
    dialog.props().onClose({} as any, 'escapeKeyDown');
    expect(props.toggleOpen).toBeCalled();
  });

  it('should have disabled submit button on both fields empty', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    const submitButton = wrapper.find(Button).at(1);
    expect(submitButton).toHaveLength(1);
    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have disabled submit button on empty last name field', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    const firstNameInput = wrapper.find(TextField).at(0);
    expect(firstNameInput).toHaveLength(1);

    firstNameInput.props().onChange({ target: { value: 'value' } } as any);

    const submitButton = wrapper.find(Button).at(1);
    expect(submitButton).toHaveLength(1);
    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have disabled submit button on empty first name field', () => {
    const wrapper = shallow(<NameChangeDialog {...props} />);
    const lastNameInput = wrapper.find(TextField).at(1);
    expect(lastNameInput).toHaveLength(1);

    lastNameInput.props().onChange({ target: { value: 'value' } } as any);

    const submitButton = wrapper.find(Button).at(1);
    expect(submitButton).toHaveLength(1);
    expect(submitButton.props().disabled).toBeTruthy();
  });

  it('should have enabledd button if all fields are filled', () => {
    const { queryByTestId } = render(<NameChangeDialog {...props} />);

    const firstNameField = queryByTestId('firstname-field');
    expect(firstNameField).toBeInTheDocument();

    const lastNameField = queryByTestId('lastname-field');
    expect(lastNameField).toBeInTheDocument();

    fireEvent.change(firstNameField, { target: { value: 'test' } });
    fireEvent.change(lastNameField, { target: { value: 'test' } });

    const submitButton = queryByTestId('submit-button');
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).not.toBeDisabled();
  });
});
