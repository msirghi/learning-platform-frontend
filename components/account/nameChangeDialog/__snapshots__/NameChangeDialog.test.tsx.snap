// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameChangeDialog component should match the snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))>
    account:nameChangeDialogTitle
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))>
      account:nameChangeDialogDescription
    </WithStyles(ForwardRef(DialogContentText))>
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      inputProps={
        Object {
          "data-testid": "firstname-field",
        }
      }
      label="account:firstName"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <div
      className="lastNameField"
    >
      <WithStyles(ForwardRef(TextField))
        fullWidth={true}
        inputProps={
          Object {
            "data-testid": "lastname-field",
          }
        }
        label="account:lastName"
        onChange={[Function]}
        value=""
        variant="outlined"
      />
    </div>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      onClick={[Function]}
    >
      account:cancelButtonLabel
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      color="primary"
      data-testid="submit-button"
      disabled={true}
      onClick={[Function]}
    >
      account:submitButtonLabel
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
